### PR TITLE
Validate order content

### PR DESCRIPTION
**Pull request by mntzrr**
Task: https://app.clickup.com/t/8694t0yt2

### CODE REVIEW

File: includes/kco-functions.php [hunk 1]
```diff
@@ -680,6 +680,51 @@ function kco_validate_order_total( $klarna_order, $order ) {
...
+		if ( 'line_item' !== $item->get_type() ) {
+			continue;
+		}
+
+		// Assume the item is a mismatch until proven otherwise. This way, in the unlikely chance of not finding any matching item, we can set the order to on-hold.
+		$mismatch = true;
+
+		// Try the specific (variation id) before the general (product id).
+		$product_id = 0 !== $item->get_variation_id() ? $item->get_variation_id() : $item->get_product_id();
+		$product    = wc_get_product( $product_id );
```
Comment (MichaelBengtsson) [inline includes/kco-functions.php:703]: The item in this case should have a method called [get_product](https://woocommerce.github.io/code-reference/classes/WC-Order-Item-Product.html#method_get_product) which should be able to replace this.

File: includes/kco-functions.php [hunk 2]
```diff
@@ -680,6 +680,51 @@ function kco_validate_order_total( $klarna_order, $order ) {
...
+		$sku        = strval( $product->get_sku() ?? $product->get_id() );
+		foreach ( $klarna_order['order_lines'] as $klarna_order_item ) {
+			if ( $klarna_order_item['reference'] === $sku ) {
+				$mismatch = $klarna_order_item['quantity'] !== $item->get_quantity();
+			}
+		}
```
Comment (MichaelBengtsson) [inline includes/kco-functions.php:709]: This does not include a check to ensure that Klarna and WooCommerce have the same order lines, only that the klarna order lines that has a matching sku in the WooCommerce order has the same quantity.

We should be checking that the order lines match, so every order line in the WooCommerce order should exist in Klarna, and all order lines that exists in Klarna should exist in WooCommerce. But keep in mind the rounding line we add in some cases, which is not added to the WooCommerce order.

This also does not have any validation for the shipping lines either, which we should add in, just make sure that we check properly for if shipping is in the iframe or not.

**Review by MichaelBengtsson — CHANGES_REQUESTED** [review]

File: includes/kco-functions.php [hunk 3]
```diff
@@ -680,6 +680,81 @@ function kco_validate_order_total( $klarna_order, $order ) {
...
+		function ( $order_line ) {
+			return 'shipping_fee' === $order_line['type'];
+		}
+	);
+	$klarna_shipping = reset( $klarna_shipping );
+
+	if ( ! empty( $klarna_shipping ) && empty( $shipping ) ) {
+		$mismatch = true;
+	} elseif ( empty( $klarna_shipping ) && ! empty( $shipping ) ) {
+		$mismatch = true;
```
Comment (MichaelBengtsson) [inline includes/kco-functions.php:733]: Would not `if ( empty( $klarna_shipping ) !== empty( $shipping ) ) {` result in the same check in the end? That way we don't have 2 separate if statements for this.

File: includes/kco-functions.php [hunk 4]
```diff
@@ -680,6 +680,81 @@ function kco_validate_order_total( $klarna_order, $order ) {
...
+	} else {
+		// If KSA is enabled, we'll skip the control.
+		$is_ksa = strpos( $shipping['reference'], 'klarna_kss' ) !== false;
+		if ( ! $is_ksa && $shipping['reference'] !== $klarna_shipping['reference'] ) {
+			$mismatch = true;
+		}
```
Comment (MichaelBengtsson) [inline includes/kco-functions.php:739]: If both `$klarna_shipping` and `$shipping` is empty here, cant we get to the case where we try to get the reference from both of the items that might not have this key causing a error?

**Review by MichaelBengtsson — CHANGES_REQUESTED** [review]

File: includes/kco-functions.php (hunk 3 above)
Comment (mntzrr) [reply includes/kco-functions.php:733]: Great idea!

**Review by MichaelBengtsson — APPROVED** [review]

**Pull request merged**